### PR TITLE
Refactor tif offset

### PR DESCRIPTION
**Pull request by xjcaa**

### CODE REVIEW

File: src/types.ts [hunk 1]
```diff
@@ -299,18 +299,25 @@ export function fromProgramOrderCompleteType(
...
-  explicitTIF?: boolean;
-  tifOffset?: number;
+  tifOptions: TIFOptions;
   orderType?: types.OrderType;
   clientOrderId?: number;
   tag?: string;
   blockhash?: string;
 }
 
+/* Only set one of these values.
```
Comment (kelvinlau20100) [inline src/types.ts:309]: old comment

File: src/types.ts [hunk 2]
```diff
@@ -299,18 +299,25 @@ export function fromProgramOrderCompleteType(
...
+ * expiryOffset - seconds in future that the order will expire.
+ */
+export interface TIFOptions {
+  expiryOffset?: number | undefined;
```
Comment (kelvinlau20100) [inline src/types.ts:313]: this should be defined if its the only one

File: src/utils.ts [hunk 3]
```diff
@@ -1646,27 +1646,28 @@ export function getProductLedger(marginAccount: MarginAccount, index: number) {
...
-  tifOffset: number
-) {
-  if (explicitTIF) {
-    return tifOffset;
+export function getTIFOffset(marketInfo: Market, tifOptions: types.TIFOptions) {
+  if (tifOptions.expiryOffset == 0) {
+    throw new Error("Invalid expiry offset");
+  }
+
+  if (tifOptions.expiryOffset == undefined) {
```
Comment (kelvinlau20100) [inline src/utils.ts:1654]: yeah im not that set on making this defined its up to you

File: src/utils.ts [hunk 4]
```diff
@@ -1646,27 +1646,28 @@ export function getProductLedger(marginAccount: MarginAccount, index: number) {
...
+    return 0;
   }
 
   let currEpochStartTs = marketInfo.serumMarket.epochStartTs.toNumber();
   let epochLength = marketInfo.serumMarket.epochLength.toNumber();
-
+  let epochEnd = currEpochStartTs + epochLength;
   let now = Exchange.clockTimestamp;
+  let desiredExpiryTs = now + tifOptions.expiryOffset;
+  let desiredOffset = desiredExpiryTs % epochLength;
```
Comment (kelvinlau20100) [inline src/utils.ts:1663]: i think we can return at this modulo?

Comment (xjcaa) [reply src/utils.ts:1663]: doesnt accomodate wrap

File: src/types.ts (hunk 2 above)
Comment (xjcaa) [reply src/types.ts:313]: discussed offline

**Pull request merged**